### PR TITLE
[SPARK-32013][SQL] Support query execution before reading DataFrame and before/after writing DataFrame over JDBC

### PR DESCRIPTION
**Pull request by moomindani**
### What changes were proposed in this pull request?

This pull request is to support query execution before reading DataFrame and after reading/writing DataFrame over JDBC.
There are two new options; `preActions` and `postActions` in DataFrame's JDBC configuration.

SQL statements specified in `preActions` will be executed before reading/writing DataFrame via JDBC.
SQL statements specified in `postActions` will be executed after writing DataFrame via JDBC.

Note: `postActions` is only supported in JDBC writer, not in JDBC reader. It is because it won't be needed so often and I was not able to find good places to implement this.

### Why are the changes needed?

For ETL workload, there is a common requirement to perform SQL statement before/after reading/writing over JDBC.
Here's examples;
- Create a view with specific conditions
- Delete/Update some records
- Truncate a table (it is already possible in `truncate` option)
- Execute stored procedure (it is also requested in SPARK-32014)

See details in the JIRA. https://issues.apache.org/jira/browse/SPARK-32013

### Does this PR introduce _any_ user-facing change?

Yes.
With this feature, users can run any SQL statements before/after reading/writing DataFrame over JDBC.

#### How to use it
Here's two examples which describes how to use the new options.

* Read
```
val df = spark.read.format("jdbc")
      .option("url", "jdbc:someUrl")
      .option("dbtable", "TEST.PEOPLE")
      .option("preActions", "delete from test.people where name = 'fred'")
      .load()
```

* Write
```
    df.write.format("jdbc")
      .option("url", "jdbc:someUrl")
      .option("dbtable", "TEST.CUSTOMQUERY")
      .option("postActions", "insert into TEST.CUSTOMQUERY values ('fred', 1)")
      .options(properties.asScala)
      .save()
```

It does not affect any existing behavior. It just adds new options to use this new feature.

### How was this patch tested?

I added test cases into `JDBCSuite.scala` and `JDBCWriteSuite.scala`, and confirmed all the tests have been passed.

### CODE REVIEW

**Comment by dilipbiswal** [comment]
@moomindani Thanks. Could we illustrate the usage of these two options via examples in the PR description ? I think, it will help the reviewers.

**Comment by moomindani** [comment]
@dilipbiswal Sure I will add it in this PR description.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala [hunk 1]
```diff
@@ -574,6 +576,41 @@ class JDBCWriteSuite extends SharedSparkSession with BeforeAndAfter {
     }
   }
 
+  test("SPARK-32013: option preActions/postActions, run SQL before writing data.") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala:579]: nit: Basically, we don't need the prefix, e.g., SPARK-32013, when adding new features.

**Comment by maropu** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -156,9 +157,16 @@ object JDBCRDD extends Logging {
     val url = options.url
     val dialect = JdbcDialects.get(url)
     val quotedColumns = requiredColumns.map(colName => dialect.quoteIdentifier(colName))
+
+    val conn = JdbcUtils.createConnectionFactory(options)
+    options.preActions match {
+      case Some(i) =>
+        runQuery(conn(), i, options)
```
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:164]: nit: `conn()` -> `conn` ?

File: docs/sql-data-sources-jdbc.md [hunk 3]
```diff
@@ -156,6 +156,20 @@ the following case-insensitive options:
      </td>
   </tr>
 
+  <tr>
+     <td><code>preActions</code></td>
+     <td>
+       You can specify custom queries which you want to run before reading data from JDBC or writing data to JDBC. It is called per DataFrame, not per session.
```
Comment (dilipbiswal) [inline docs/sql-data-sources-jdbc.md:162]: @moomindani Do we allow multiple pre actions ? Or we allow just one ?

File: docs/sql-data-sources-jdbc.md [hunk 4]
```diff
@@ -156,6 +156,20 @@ the following case-insensitive options:
...
+     </td>
+  </tr>
+
+  <tr>
+     <td><code>postActions</code></td>
+     <td>
+        This is a JDBC writer related option. You can specify custom queries which you want to run after writing data to JDBC. It is called per DataFrame, not per session.
```
Comment (dilipbiswal) [inline docs/sql-data-sources-jdbc.md:169]: @moomindani Same question for post action.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 5]
```diff
@@ -122,6 +122,19 @@ object JdbcUtils extends Logging {
     }
   }
 
+  /**
+   * Runs a custom query against a table from the JDBC database.
+   */
+  def runQuery(conn: Connection, queryString: String, options: JDBCOptions): Unit = {
+    val statement = conn.prepareStatement(queryString)
```
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:129]: @moomindani Are we preventing user's to pass in "select queries" as part of pre or post actions ? If they pass in a query, then would we get an result set when we call "stme.execute()" that we need to free ? 

Another question is , what happens when a post action fails ? We don't have a notion of a transaction here, right ? In that case we probably need to document our behavior ?

**Comment by maropu** [comment]
> For ETL workload, there is a common requirement to perform SQL statement before/after reading/writing over JDBC.
Here's examples;
Create a view with specific conditions
Delete/Update some records
Truncate a table (it is already possible in truncate option)
Execute stored procedure (it is also requested in SPARK-32014)

@gatorsmile WDYT the usecases described above? I'm not sure it's reasonable or not because I've never met users making such a request.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala (hunk 1 above)
Comment (moomindani) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala]: Removed the prefix.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: `scanTable` returns `JDBCRDD` with `JdbcUtils.createConnectionFactory(options)`, not with `JdbcUtils.createConnectionFactory(options)()`.

BTW `scanTable` is called per RDD so I moved the implementation to `resolveTable`. Here, I used `conn` as the argument for `runQuery`.

File: docs/sql-data-sources-jdbc.md (hunk 3 above)
Comment (moomindani) [reply docs/sql-data-sources-jdbc.md]: In the initial pull-request, we only supported only one query. However, I added support for multiple queries separated by semicolon.

File: docs/sql-data-sources-jdbc.md (hunk 4 above)
Comment (moomindani) [reply docs/sql-data-sources-jdbc.md]: In the initial pull-request, we only supported only one query. However, I added support for multiple queries separated by semicolon.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 5 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: No, we do not prevent "select queries". Users can run any of "select", "insert", "update", "delete", etc.
Right, your concern is correct. I added implementation to close ResultSet for "select queries".

In addition, I added transaction and error handling logic. I also added unit tests for exceptions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (kiszk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Does this call `conn` have a side-effect? According to [this document](https://github.com/databricks/scala-style-guide#parentheses), `()` means that it has a side-effect.

IMHO, since other places use `conn`, I guess that this call does not have a side-effect.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 6]
```diff
@@ -122,6 +122,37 @@ object JdbcUtils extends Logging {
     }
   }
 
+  /**
+   * Runs a custom query against a table from the JDBC database.
+   */
+  def runQuery(conn: Connection, actions: String, options: JDBCOptions): Unit = {
+    val autoCommit = conn.getAutoCommit
+    conn.setAutoCommit(false)
+    val queries = actions.split(";")
```
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:131]: @moomindani We probably need to parse it smartly if we want to allow multi statements. Above may not work very well if the queries have embedded semi colons in a string literal ? Also, not sure if some databases allow semi colons in the identifier like table names.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 7]
```diff
@@ -122,6 +122,37 @@ object JdbcUtils extends Logging {
     }
   }
 
+  /**
+   * Runs a custom query against a table from the JDBC database.
+   */
+  def runQuery(conn: Connection, actions: String, options: JDBCOptions): Unit = {
+    val autoCommit = conn.getAutoCommit
+    conn.setAutoCommit(false)
```
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:130]: @moomindani Typically in my experience, in traditional databases, we do the pre and post trigger actions in the same transaction as the main action. I was saying that in spark, it's probably not possible as the main action could be done in a distributed fashion. So if we do this, we probably have to document it telling our users that, when a post action fails, we still can have the pre and post action committed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 8]
```diff
@@ -122,6 +122,37 @@ object JdbcUtils extends Logging {
...
+    val queries = actions.split(";")
+    try {
+      queries.foreach { query =>
+        val queryString = query.trim()
+        val statement = conn.prepareStatement(queryString)
+        try {
+          statement.setQueryTimeout(options.queryTimeout)
+          val hasResultSet = statement.execute()
```
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:138]: @moomindani Is there a reason to allow a query here ? I am asking, since no one can consume this result set ? Or we are allowing it, just coz we are not able to parse the input to know what kind of statement it is ?

**Comment by dilipbiswal** [comment]
@moomindani Thanks a lot for working on it. Would be nice to wait on answer to @maropu's question.

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:138]: It is to support stored procedures. (related to https://issues.apache.org/jira/browse/SPARK-32014)
For example, MySQL does not return result sets in stored procedures. In such case, we can use `executeUpdate()`.
However, some other databases returns result sets in stored procedures. To support such use-case, we need to consider both `executeQuery()` and `executeUpdate()`. That's the reason why I used `execute()` here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 7 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:130]: Makes sense. I have updated the document to include this point.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Thanks for pointing the reference. It makes sense.
BTW, in the current diff, all the `runQuery` calls have `conn` instead of `conn()`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 6 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:131]: How about reusing the existing code? https://github.com/apache/spark/blob/master/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala#L521-L578

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (kiszk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Sure. Now, they are fine.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 6 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Thank you, it makes sense to me.
I modified it to use the linked code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:140]: @moomindani I see. But if we call a cursory stored procedure in the pre/post action, what do we do with the result set retuned from it ? I am thinking users would call a cursory stored procedure if they want to do something with the results, no ? There are also some engines that can return multiple result set from a stored procedure.

https://stackoverflow.com/questions/53621106/db2-multiple-result-set-stored-procedure

**Comment by dilipbiswal** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:140]: Currently result sets are always ignored if `preActions` or `postActions` return them. As you said, cursor won't be able to be used for reading as DataFrame.

By design, basically DDL queries are expected for `preActions` and `postActions`.
The reason why I used `execute` instead of `executeUpdate` is just to avoid exceptions due to stored procedure which returns result sets. Not to use result sets as an input for DataFrame.

Do you think it is enough for us to explain that result sets are always ignored in `preActions` and `postActions` if query results the result sets?
Or should we narrow down use-cases to support only DDL queries (which do not return result sets) by using `executeUpdate` instead of `execute`?

My idea was the first one, but I want to follow community's opinion.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:140]: @moomindani 
IMHO, the best behavior would be to only allow DDL or DML (insert/update/delete) in pre and post actions. But if its not achievable, then we have to clearly document what is allowed and ramifications of users specifying other kind of statements. Lets please get input from @maropu and @HyukjinKwon to decide on the best course to move forward.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: I really appreciate your kind review. I had a quick conversation with @maropu and he agreed with your idea. 
After reviewing your comment, I changed my mind to allow only DDL and DML (insert/update/delete) queries in `preActions/postActions` to use `addBatch()` and `executeBatch()` instead of looping `execute()`. I reflected changes in the codes and also updated docs.

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: @moomindani The changes look reasonable to me. One small question. Read the following :
```
JDBC drivers are not required to support this feature. You should use the 
DatabaseMetaData.supportsBatchUpdates() method to determine if the
target database supports batch update processing. The method returns 
true if your JDBC driver supports this feature.
```
at [link](https://www.tutorialspoint.com/jdbc/jdbc-batch-processing.htm)

So can we assume all the underlying JDBC implementations we have support this
feature ? As a start, we could just check DatabaseMetaData.supportsBatchUpdates() and 
raise an error saying pre and post actions are not supported ? Or we could address it as
a follow-up ?

Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Your concern is valid, currently most of JDBC drivers (including MySQL, PostgreSQL, Oracle, SQL Server, Redshift, Athena, Impala, Snowflake, etc.) support batchUpdate, but I guess not all drivers support it.
My current implementation followed existing codes in `JdbcUtils.savePartition()` since it already used `addBatch()` and `executeBatch()`.

I could come up with following 3 possible directions for it.
(a) Keep current implementation. The `preActions`/`postActions` are only supported in JDBC drivers which support batchUpdates.
(b) Loop `executeUpdate()` instead of using `addBatch()`/`executeBatch()`. Although it sacrifices performance benefit, it will be safer when JDBC drivers do not support batchUpdates.
(c) Add check logic to use `DatabaseMetaData.supportsBatchUpdates()`. 
  - (c-1). Add it only to `preActions`/`postActions`
  - (c-2). For consistency, add it into both `preActions`/`postActions` and `JdbcUtils.savePartition()`.

Personally I prefer (a) or (b). 
If we do (c), we might need to implement two different ways to send queries.

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: @moomindani Option `(a)` sounds okay to me as its simpler. We can always improve the behavior later. 
wdyt @maropu  ?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yea, `(a)` sounds fine to me.

> Only DDL or DML (insert/update/delete) are allowed. 

Btw, how do we control this? what if SELECT queries given?

**Comment by kiszk** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: @maropu 
My expectation is that the JDBC driver would reject a statement other than DML (insert/update/delete). Looking at this link https://stackoverflow.com/questions/24577358/how-to-add-select-statements-in-addbatch, it seems like executeBatch may also reject DDLS.

@moomindani Can you please confirm if we can execute DDL's in executeBatch ?

**Comment by dilipbiswal** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 9]
```diff
@@ -156,6 +164,7 @@ object JDBCRDD extends Logging {
     val url = options.url
     val dialect = JdbcDialects.get(url)
     val quotedColumns = requiredColumns.map(colName => dialect.quoteIdentifier(colName))
+
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:167]: nit: revert the unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala [hunk 10]
```diff
@@ -46,7 +46,18 @@ class JdbcRelationProvider extends CreatableRelationProvider
...
     val conn = JdbcUtils.createConnectionFactory(options)()
+
+    var parametersWithoutPreActions: Map[String, String] = parameters
     try {
+      options.preActions match {
+        case Some(i) =>
+          runQuery(conn, i, options)
+
+          // Remove preActions to avoid duplicate execution when writing data
+          parametersWithoutPreActions = parameters.-(JDBCOptions.JDBC_PRE_ACTIONS_STRING)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala:57]: The code looks weird. What's the case you're worried about here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala [hunk 11]
```diff
@@ -206,6 +209,9 @@ class JDBCOptions(
   }
   // The principal name of user's keytab file
   val principal = parameters.getOrElse(JDBC_PRINCIPAL, null)
+
+  // the query used after writing data
+  var postActions = parameters.get(JDBC_POST_ACTIONS_STRING)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:214]: var -> val?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala [hunk 12]
```diff
@@ -116,6 +116,9 @@ class JDBCOptions(
   // number of seconds. Zero means there is no limit.
   val queryTimeout = parameters.getOrElse(JDBC_QUERY_TIMEOUT, "0").toInt
 
+  // the query used before reading/writing data
+  var preActions = parameters.get(JDBC_PRE_ACTIONS_STRING)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala:120]: var -> val?

File: docs/sql-data-sources-jdbc.md [hunk 13]
```diff
@@ -156,6 +156,20 @@ the following case-insensitive options:
      </td>
   </tr>
 
+  <tr>
+     <td><code>preActions</code></td>
+     <td>
+       Custom queries which you want to run before reading data from JDBC or writing data to JDBC. Only DDL or DML (insert/update/delete) are allowed. It is called per DataFrame, not per session. You can specify multiple queries separated by semicolon. When exceptions occur in preActions, the queries in preActions will be rollbacked.
```
Comment (maropu) [inline docs/sql-data-sources-jdbc.md:162]: >> It is called per DataFrame, not per session.

I think this statement looks ambiguous. What does `per DataFrame` mean? Probably, `per session` means `sessionInitStatement` though.

One more question; is there any usecase `preActions` can handle and `sessionInitStatement` cannot?

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryStatementUtils.scala [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * Util functions for handling Query statements.
+ */
+object QueryStatementUtils {
+
+  // Adapted splitSemiColon from Hive 2.3's CliDriver.splitSemiColon.
+  // Note: [SPARK-31595] if there is a `'` in a double quoted string, or a `"` in a single quoted
+  // string, the origin implementation from Hive will not drop the trailing semicolon as expected,
+  // hence we refined this function a little bit.
+  def splitSemiColon(line: String): JList[String] = {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/util/QueryStatementUtils.scala:31]: How about moving this helper func into `org.apache.spark.sql.catalyst.util.StringUtils`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 15]
```diff
@@ -122,6 +123,33 @@ object JdbcUtils extends Logging {
...
+   */
+  def runQuery(conn: Connection, actions: String, options: JDBCOptions): Unit = {
+    val autoCommit = conn.getAutoCommit
+    conn.setAutoCommit(false)
+
+    val commands = QueryStatementUtils.splitSemiColon(actions).asScala
+    val statement = conn.createStatement()
+    statement.setQueryTimeout(options.queryTimeout)
+
+    commands.foreach { command =>
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:137]: How about adding `logInfo` here? https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala#L283

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 8 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: > Btw, how do we control this? what if SELECT queries given?

I confirmed that it returns exception like `Method is not allowed for a query`.

> Can you please confirm if we can execute DDL's in executeBatch ?

Sure, let me confirm it and add test to use DDL.

Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: I added some tests for DDLs. As far as I tested, `addBatch()`/`executeBatch()` worked also for DDLs.
It seems it also works for MySQL Connector/J since `executeBatch` just calls `executeUpdate` internally. https://github.com/mysql/mysql-connector-j/blob/dbd865c9fc68701efc70197ea1e65002520bed91/src/main/user-impl/java/com/mysql/cj/jdbc/StatementImpl.java#L859
I found a blog post which used `addBatch()`/`executeBatch()` with MySQL. https://www.roseindia.net/tutorial/java/jdbc/jdbcbatch/jdbcbatchdatabasetablecreation.html

However, I could not find any clear documentation which says DDLs are supported in `executeBatch()`. What I could find is that DDLs are supported in `executeUpdate()` and `executeLargeUpdate()`.
https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/Statement.html

Should we move from (a) to (b) as safer approach?

File: docs/sql-data-sources-jdbc.md (hunk 13 above)
Comment (maropu) [reply docs/sql-data-sources-jdbc.md:162]: > ~One more question; is there any usecase `preActions` can handle and `sessionInitStatement` cannot?~

Ah, I see. `sessionInitStatement` is executed on executor-sides and It seems `preActions` is proposed to be executed in a driver side. Probably, you'd be better to describe what's a difference between them in the doc clearly.

**Comment by maropu** [comment]
btw, `postActions` is only for write paths? The title does not say so...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 9 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Reverted.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala (hunk 11 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala]: Modified it to use `val`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala (hunk 12 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala]: Modified it to use `val`.

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryStatementUtils.scala (hunk 14 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryStatementUtils.scala]: Moved this logic to `org.apache.spark.sql.catalyst.util.StringUtils` as suggested.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 15 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:136]: Added `logInfo` in similar way.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala (hunk 10 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala:74]: As I added the comments, it is for preventing duplicate execution of `preActions` in write path.
There are two methods around here.
a) `createRelation(SQLContext, SaveMode, Map[String, String], DataFrame)`
b) `createRelation(SQLContext, Map[String, String])`

(a) is called only for writes, but (b) is called for both reads and writes
To avoid duplicate execution of preActions in writes, we need to remove preActions parameter here before passing it to (b)

Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcRelationProvider.scala:74]: As I added the comments, it is for preventing duplicate execution of preActions in write path.
There are two methods around here.
a) `createRelation(SQLContext, SaveMode, Map[String, String], DataFrame)`
b) `createRelation(SQLContext, Map[String, String])`

(a) is called only for writes, but (b) is called for both reads and writes
To avoid duplicate execution of preActions in writes, we need to remove preActions parameter here before passing it to (b)

File: docs/sql-data-sources-jdbc.md (hunk 13 above)
Comment (moomindani) [reply docs/sql-data-sources-jdbc.md]: You are right.. this logic should not be implemented here.

I moved it to `createRelation` to make sure that it is called only once and to have consistency in the timing for both read path and write path.
I also updated the document to describe the differences.

**Comment by moomindani** [comment]
> btw, `postActions` is only for write paths? The title does not say so...

Right. I modified the title.

**Comment by maropu** [comment]
retest this please

**Comment by moomindani** [comment]
retest this please

**Comment by moomindani** [comment]
retest this please

**Comment by moomindani** [comment]
retest this please

**Comment by kiszk** [comment]
@moomindani now, the conflict happens. Could you please resolve it?

**Comment by moomindani** [comment]
I resolved recent conflicts.
Folks, can you review this and proceed it? I thought there are no blockers now.

**Review by gatorsmile — CHANGES_REQUESTED** [review]
This is not JDBC specific. All the data sources might require pre/post execution hooks/triggers/actions. We need to consider this request in the API design of DSV2. 

cc @cloud-fan

**Comment by moomindani** [comment]
@gatorsmile Thank you for your comment.
I understand that this kind of hooks are not specific to JDBC, it can be general to all the data sources.
However, for JDBC, it should be executed in data source side. This part can be specific to JDBC (or other data sources which include execution engine.).

Of course we can have discussion in DSV2 API design, however, can we proceed this PR separately?

**Comment by moomindani** [comment]
@gatorsmile Just a reminder.. Can you take a look?

**Comment by cloud-fan** [comment]
This sounds like JDBC specific, as we can't use SQL to define actions for data sources like parquet.

To clarify your use case, do you really need hooks, or just need the ability to run arbitrary SQL against the JDBC server with Spark?

**Comment by moomindani** [comment]
@cloud-fan We need hooks right before/afterJDBC reads/writes especially for ETL purpose. It is also frequently asked request from our customers.

**Comment by abhishekboga** [comment]
Hello everyone, I know this pr has been closed but it would be really helpful if I get a resolution on the same topic. So here I am trying to call snowflake procedure in the aws glue job using Pyspark. Below is the code:
df = spark.read.format(SNOWFLAKE_SOURCE_NAME).options(**sfOptions).option("query", "call SP_UTILIZATION_INSIGHTS_LOOKUP_TABLE()").option("preActions", "call SP_UTILIZATION_INSIGHTS_LOOKUP_TABLE()").load()

While execution I'm getting an error as SQL Compilation error: at 'call'.

In the log files what I found is it basically creates a select statement as 'Select * from (call SP_UTILIZATION_INSIGHTS_LOOKUP_TABLE()) where 1=0' and due to this it failed.

Any help here?

**Comment by cloud-fan** [comment]
@peter-toth I vaguely remembered that you added a new JDBC option to support this case?

**Comment by abhishekboga** [comment]
@peter-toth , can you share some examples of sessionInitStatement to call the stored procedure?

**Comment by peter-toth** [comment]
Can you please try:
```
df = spark.read.format(SNOWFLAKE_SOURCE_NAME)
  .option("sessionInitStatement", "call SP_UTILIZATION_INSIGHTS_LOOKUP_TABLE()")
  .option("query", "SELECT * FROM your_table_that_is_populated")
  .load()
```

**Comment by abhishekboga** [comment]
@peter-toth , I have tried your example but it doesn't call the SP which populates a table. What I was expecting was SP will populate the table and then that table which is used in the "query" but it didn't happen.

**Comment by abhishekboga** [comment]
@peter-toth, any update on the above query? It's not executing the SP.

**Comment by peter-toth** [comment]
@abhishekboga, sorry I can't test this feature with Snowflake, but I added the following simple unittest using H2 into `sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala`: 
```
  test("SPARK-32013: option sessionInitStatement, call stored function") {
    val initSQL =
      """CREATE ALIAS NEXT_PRIME AS '
        |String nextPrime(String value) {
        |  return new BigInteger(value).nextProbablePrime().toString();
        |}';
        |INSERT INTO TEST.seq VALUES(NEXT_PRIME(1000))
        |""".stripMargin
    val df = spark.read.format("jdbc")
      .option("url", urlWithUserAndPass)
      .option("query", "SELECT * FROM TEST.seq WHERE ID > 1000")
      .option("sessionInitStatement", initSQL)
      .load()
    assert(df.collect() === Array(Row(1009)))
  }
```
and it does work.
